Time-Card/TOOL/main.go

```diff
@@ -0,0 +1,146 @@
+	ImageSize        uint32
+	HardwareRevision uint16
+	CRC              uint16
+}
+
+func main() {
+	c := &Config{}
+
+	var logLevel string
+	flag.BoolVar(&c.Apply, "apply", false, "Create new firmware file with the header in the beginning")
+	flag.StringVar(&c.InputPath, "input", "", "Path to raw firmware file")
+	flag.StringVar(&c.OutputPath, "output", "", "Path to firmware file with header (will be overwritten")
+	flag.IntVar(&c.VendorId, "vendor", 0, "PCI VEN_ID to add to header")
+	flag.IntVar(&c.DeviceId, "device", 0, "PCI DEV_ID to add to header")
+	flag.IntVar(&c.HardwareRevision, "hw", 0, "PCI REV_ID (Hardware revision) to add to header")
+	flag.StringVar(&logLevel, "loglevel", "info", "Set a log level. Can be: debug, info, warning, error")
+	flag.Parse()
+
+	switch logLevel {
+	case "debug":
+		log.SetLevel(log.DebugLevel)
+	case "info":
+		log.SetLevel(log.InfoLevel)
+	case "warning":
+		log.SetLevel(log.WarnLevel)
+	case "error":
+		log.SetLevel(log.ErrorLevel)
+	default:
+		log.Fatalf("Unrecognized log level: %v", logLevel)
+	}
+
+	if c.InputPath == "" {
+		log.Fatalf("No input file provided")
+	}
+
+	if c.Apply && c.OutputPath == "" {
+		log.Fatalf("No output file provided")
+	}
+
+	if c.VendorId <= 0 || c.VendorId > 65535 {
+		log.Fatalf("Empty or incorrect PCI Vendor ID specified")
+	}
+
+	if c.DeviceId <= 0 || c.DeviceId > 65535 {
+		log.Fatalf("Empty or incorrect PCI Device ID specified")
+	}
+
+	if c.HardwareRevision < 0 || c.HardwareRevision > 65535 {
+		log.Fatalf("Incorrect PCI Device Revision ID specified")
+	}
+
+	hdr := &Header{
+		Magic:            [4]byte{'O', 'C', 'T', 'C'},
+		VendorId:         uint16(c.VendorId),
+		DeviceId:         uint16(c.DeviceId),
+		HardwareRevision: uint16(c.HardwareRevision),
+	}
+
+	inputFile, err := os.Open(c.InputPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer inputFile.Close()
+	stat, err := inputFile.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	hdr.ImageSize = uint32(stat.Size())
+
+	var outputFile *os.File
+	if c.Apply {
+		outputFile, err = os.Create(c.OutputPath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		h := new(bytes.Buffer)
+		binary.Write(h, binary.BigEndian, &hdr)
+		_, err = outputFile.Write(h.Bytes())
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	crcTable := crc.MakeTable(crc.CRC16_ARC)
+	buf := make([]byte, 16384)
+	crc16 := uint16(0)
+
+
+	for n, err := inputFile.Read(buf); n > 0 && err == nil; n, err = inputFile.Read(buf) {
+		crc16 = crc.Update(crc16, buf[:n], crcTable)
+		if c.Apply {
+			_, err = outputFile.Write(buf[:n])
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	crc16 = crc.Complete(crc16, crcTable)
+
+	hdr.CRC = crc16
+
+	if c.Apply {
+		h := new(bytes.Buffer)
+		binary.Write(h, binary.BigEndian, &hdr)
+		_, err := outputFile.WriteAt(h.Bytes(), 0)
+		if err != nil {
+			log.Fatal(err)
+		}
+		outputFile.Close()
+		fmt.Printf("Firmware image with header is written to %s\n", c.OutputPath)
+	}
```

Comment:
This should go to another function (`crc or whatever) and have separate tests set.